### PR TITLE
ListView Android performance refactor

### LINKED ISSUE #2857: Performance issues in RecyclerView based ListView
**Describe the bug**

[From the forum](https://community.appinventor.mit.edu/t/listview-freezes-after-last-ai-update/40144?u=ewpatton): The RecyclerView based ListView seems to have performance degradation for large lists.

**Affects**

- [ ] Designer
- [ ] #3017
- [x] Companion
- [x] Compiled apps
- [ ] Buildserver
- [ ] Debugging
- [ ] Other... (please describe)

**Expected behavior**

Switching to the RecyclerView should have had better performance than the default ListView implementation. We need to investigate this regression.

**Steps to reproduce**

1. Create a ListView and populate it with a few thousand elements.

### PR DESCRIPTION
**Pull request by SusanRatiLane**
General items:
- [x] `ant tests` passes on my machine

- [X] I branched from `ucr`
- [X] My pull request has `ucr` as the base

Further, if you've changed the blocks language or another user-facing designer/blocks API (added a SimpleProperty, etc.):

- [ ] I have updated the corresponding version number in YaVersion.java
- [ ] I have updated the corresponding upgrader in YoungAndroidFormUpgrader.java (components only)
- [ ] I have updated the corresponding entries in versioning.js

^^ I have actually never done this, so we'll need to talk.

*Description*

This is a performance refactor submitted by @patryk84a in February 2024. He also included some methods for adding and removing items that the class should probably have. I included implementations for these in iOS. 

He also included one aesthetic property for which I included a stub for iOS.

Fixes #2857 

Original PR (now closed): #3099 

Co-author: @patryk84a

### CODE REVIEW

**Comment by SusanRatiLane** [comment]
Hmm. I thought the tests were passing, but two ListView ones are not. Patryk's changes have broken an assumption the tests are making, but I'm trying to sort out what it is.

**Comment by patryk84a** [comment]
The HintSearchText property has been removed. Maybe instead for Android add a fixed hint text but from the Android resources "@android:string/search_go", then we should get the string in the language set in the phone, without the possibility of defining your own, which would be compatible with IOS

**Comment by SusanRatiLane** [comment]
> The HintSearchText property has been removed. Maybe instead for Android add a fixed hint text but from the Android resources "@android:string/search_go", then we should get the string in the language set in the phone, without the possibility of defining your own, which would be compatible with IOS

Actually, that was removed because someone else wrote a PR that adds the property for both Android and iOS. I should probably have mentioned that in the PR description.

File: appinventor/components/src/com/google/appinventor/components/common/ListOrientation.java
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/common/ListOrientation.java:1]: Line endings.

File: appinventor/components/src/com/google/appinventor/components/runtime/ListBounceEdgeEffectFactory.java
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/ListBounceEdgeEffectFactory.java:1]: Line endings need to be switched to Unix. There are also a few indentation/wrapping issues that IntelliJ should be able to help with.

**Review by ewpatton — CHANGES_REQUESTED** [review]
In addition to the file level and inline comments, the LISTVIEW_COMPONENT_VERSION value in YaVersion needs to be bumped to the change in the properties, and the appropriate code paths in YoungAndroidFormUpgrader.java and versioning.js need to be added.

**Comment by patryk84a** [comment]
I guess I can't make any changes here, right?

**Comment by SusanRatiLane** [comment]
> I guess I can't make any changes here, right?

I should be able to give you access. I'll take a look.

File: appinventor/components/src/com/google/appinventor/components/common/LayoutType.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  @Default
+  MainText(0),
+  MainText_DetailText_Vertical(1),
+  MainText_DetailText_Horizontal(2),
+  Image_MainText(3),
+  Image_MainText_DetailText_Vertical(4);
+
+  private final int layout;
+
+   LayoutType(int value) {
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/components/src/com/google/appinventor/components/common/LayoutType.java:24]: > This seems like it should be private. The indentation is also off by 1.

IntelliJ informs me that private is redundant for enum constructions.

File: appinventor/components/src/com/google/appinventor/components/common/ListOrientation.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * Defines a parameter of ListOrientation used by the ListView component.
+ */
+public enum ListOrientation implements OptionList<Integer> {
+  @Default
+  Vertical(0),
+  Horizontal(1);
+
+  private final int orientation;
+
+  ListOrientation(int value) {
```
[the remark this answers is not in the document]
Comment (SusanRatiLane) [reply appinventor/components/src/com/google/appinventor/components/common/ListOrientation.java]: Ditto for redundant.

**Comment by patryk84a** [comment]
Fix duplicate call to notifyDataSetChanged - I don't really understand this fix because I didn't find a duplicate call to the notifyDataSetChanged() method. How is it supposed to help to place this method in several places instead of in the clearSelection() function, since this function is called anyway. 

Any syntax errors, whitespace, etc. were caused by working in an editor without intelij. I think my current code editor would catch those errors now. It's great that there are public methods related to Multiselect. We'll be able to use it in the extension. 

I tried to reflect the appearance of the new list view in the Mock file in gwt, but I can't make these components behave the way I want them to. The problem is probably with automatic dimensioning, e.g. to the width of the parent, etc. In gwt, I think you have to give everything in constant calculated dimensions.

**Comment by patryk84a** [comment]
Ok, I tested it and it seems to work fine. There is one small detail that I didn't notice before. There is no margin between the image and the text. The text is glued directly to the image. I think a quick fix is ​​to add a margin to the right of the image, then it will work for every layout.

layoutParamsImage.setMargins(0, 0, 15, 0);

The padding for the element is set to 15, so I think the margin can also be set to 15 so there will be equal space around the image.

**Comment by jisqyv** [comment]
I just pushed up the squashed version, after merging in ucr

**Review by jisqyv — CHANGES_REQUESTED** [review]
So, Everything appears to work in the Companion, but if I compile an App, on my Samsung phone running Android 14, the filter bar does not show up!

**Comment by jisqyv** [comment]
Additional info. The filter bar *is* there, but it has a black background, while with the companion, the background was white.

**Comment by jisqyv** [comment]
I'm removing my objections. The behavior I'm seeing (black search bar) is also present in our current production system, so this is not a new bug.

**Review by jisqyv — APPROVED** [review]
LGTM

**Review by ewpatton — APPROVED** [review]
LGTM

**Comment by ewpatton** [comment]
Note that I had previously fixed those comments in the individual changes prior the squash.

**Comment by patryk84a** [comment]
Is it possible to make the small correction I mentioned before the release, i.e. the margin on the right side of the image? Now the text "touches" the image in an unsightly way. I also mentioned the black background of the search bar in the community thread, apparently no one reads the comments posted there, but this can be left for the next ListView correction.

**Comment by patryk84a** [comment]
> Additional info. The filter bar _is_ there, but it has a black background, while with the companion, the background was white.

Yes, I've noticed that too. The theme type auto-detection doesn't work very well. One idea was to set the EditText background to transparent so that it takes the ListView background, and set the search text color to the MainText color. If the app author sets the ListView colors to the appropriate ones, the search bar will also have the appropriate colors. We don't have to decide the colors for the app author.

**Pull request merged**